typing/printtyp.ml

```diff
@@ -225,7 +231,28 @@ let enabled = ref true
 let enable b = enabled := b
 
 (* Names bound in recursive definitions should be considered as bound
-   in the environment when printing identifiers *)
+   in the environment when printing identifiers but not when trying
+   to find shortest path.
+   For instance, if we define
+   [{
+   module Avoid__me = struct
+     type t = A
+   end
+   type t = X
+   type u = t * t
+   module M = struct
+     type t = A of u
+     type r = Avoid__me.t
+   end
+  }]
+  It is is important that in the definition of [t] that the outer type [t] is
+  printed as [t/2] reserving the name [t] to the type being defined in the
+  current recursive definition.
+     Contrarily, in the definition of [r], one should not shorten the
+  path [Avoid__me.t] to [r] until the end of the definition of [r].
+  The [bound_in_recursion] bridges the gap between those two slightly different
+  notions of printing environment.
```

Comment:
If I understand correctly, we get a problem if, working on the occurrence of `Avoid__me.t` in the definition of `M.r`, we would call short-path with an environment where `r` is already in scope and bound to `Avoid__me.t`. Naive idea: could we use an environment where `r` is bound but abstract? Would this allow to use a single notion of environment for both use-cases?

Generally the idea is that in `let rec x = t in u`, you would use for `t` an environment with `x` is bound but abstract, and you would use the full definition `x = t` only in `u`.

(This assumes there is a notion of "abstract" for each kind of objects that can be recursively defined and are shown by printtyp. I was going to make a joke on abstract module types saving the day again, but I don't know how to define a recursive module type anyway!)

Reply:
That might work with the current implementation of `short-path`.

Reply:
Who cares about the other implementation? (If someone did, they would submit it upstream, right?)

Reply:
I am pretty sure that I am part of the set of people caring. And replacing the current algorithm would be a good time to remove the friction between `short-path` and the rest of `Printtyp`.